File: sdks/sms8-android/sms8/src/main/java/io/sms8/SMS8.kt
package io.sms8

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL

/**
 * SMS8 Android SDK — send and verify SMS OTP codes through your own Android phone.
 *
 * No Twilio, no Vonage Verify, no MessageBird. No per-OTP fees, no markups,
 * no A2P 10DLC. Free 5-day trial at https://sms8.io.
 *
 * ```kotlin
 * val sms8 = SMS8(apiKey = BuildConfig.SMS8_API_KEY)
 *
 * lifecycleScope.launch {
 *     val sent = sms8.sendOTP(phone = "+14155550100")
 *     // show OTP UI...
 *     val result = sms8.verifyOTP(phone = "+14155550100", code = "482937")
 *     if (result.verified == true) { /* allow login */ }
 * }
 * ```
 */
class SMS8(
    private val apiKey: String,
    private val baseUrl: String = "https://mcp.sms8.io",
) {
    init {
        require(apiKey.isNotBlank()) { "SMS8: apiKey must not be blank. Get one at https://sms8.io" }
    }

    data class SendOTPResult(
        val success: Boolean,
        val otpId: Int?,
        val phone: String?,
        val expiresAt: String?,
        val expiresIn: Int?,
        val error: String?,
    )

    data class VerifyOTPResult(
        val success: Boolean,
        val verified: Boolean?,
        val error: String?,
        val attemptsLeft: Int?,
    )

    // ── SMS ────────────────────────────────────────────────────────────
    suspend fun sendSMS(
        phone: String,
        message: String,
        deviceId: Int? = null,
        simSlot: String? = null,
    ): JSONObject = mcpCall("send_sms", JSONObject().apply {
        put("phone", phone)
        put("message", message)
        deviceId?.let { put("device_id", it) }
        simSlot?.let  { put("sim_slot",  it) }
    })

    // ── OTP send ───────────────────────────────────────────────────────
    suspend fun sendOTP(
        phone: String,
        length: Int = 6,
        template: String? = null,
        expiresIn: Int = 300,
        maxAttempts: Int = 5,
        deviceId: Int? = null,
        simSlot: String? = null,
    ): SendOTPResult {
        val args = JSONObject().apply {
            put("phone", phone)
            put("length", length)
            put("expires_in", expiresIn)
            put("max_attempts", maxAttempts)
            template?.let { put("template", it) }
            deviceId?.let  { put("device_id", it) }
            simSlot?.let   { put("sim_slot",  it) }
        }
        val out = mcpCall("send_otp", args)
        return SendOTPResult(
            success    = out.optBoolean("success", false),
            otpId      = out.optInt("otp_id").takeIf { out.has("otp_id") },
            phone      = out.optString("phone").takeIf { it.isNotEmpty() },
            expiresAt  = out.optString("expires_at").takeIf { it.isNotEmpty() },
            expiresIn  = out.optInt("expires_in").takeIf { out.has("expires_in") },
            error      = out.optString("error").takeIf { it.isNotEmpty() },
        )
    }

    // ── OTP verify ─────────────────────────────────────────────────────
    suspend fun verifyOTP(phone: String, code: String): VerifyOTPResult {
        val out = mcpCall("verify_otp", JSONObject().apply {
            put("phone", phone)
            put("code",  code)
        })
        return VerifyOTPResult(
            success      = out.optBoolean("success", false),
            verified     = out.optBoolean("verified").takeIf { out.has("verified") },
            error        = out.optString("error").takeIf { it.isNotEmpty() },
            attemptsLeft = out.optInt("attempts_left").takeIf { out.has("attempts_left") },
        )
    }

    // ── HTTP plumbing ──────────────────────────────────────────────────
    private suspend fun mcpCall(tool: String, args: JSONObject): JSONObject = withContext(Dispatchers.IO) {
        args.put("api_key", apiKey)
        val body = JSONObject().apply {
            put("jsonrpc", "2.0")
            put("method",  "tools/call")
            put("params",  JSONObject().apply {
                put("name", tool)
                put("arguments", args)
            })
            put("id", System.currentTimeMillis())
        }
        val url  = URL(baseUrl)
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            doOutput = true
            setRequestProperty("Content-Type", "application/json")
            setRequestProperty("Authorization", "Bearer $apiKey")
            setRequestProperty("User-Agent", "sms8-android/1.0.0")
            connectTimeout = 15_000
            readTimeout    = 30_000
        }
        try {
            conn.outputStream.use { it.write(body.toString().toByteArray()) }
            val code = conn.responseCode
            if (code !in 200..299) {
                val err = conn.errorStream?.bufferedReader()?.use { it.readText() } ?: ""
                throw IOException("HTTP $code: $err")
            }
            val response = conn.inputStream.bufferedReader().use { it.readText() }
            val env = JSONObject(response)
            env.optJSONObject("error")?.let {
                throw IOException(it.optString("message", "MCP error"))
            }
            val content = env.optJSONObject("result")?.optJSONArray("content") ?: JSONArray()
            if (content.length() == 0) return@withContext env.optJSONObject("result") ?: JSONObject()
            val text = content.getJSONObject(0).optString("text", "")
            return@withContext try { JSONObject(text) } catch (_: Exception) { JSONObject().apply { put("text", text) } }
        } finally {
            conn.disconnect()
        }
    }
}

File: sdks/sms8-swift/Sources/SMS8/SMS8.swift
//
//  SMS8.swift
//  SMS8 — Swift SDK for sending and verifying SMS OTP codes via your own phone.
//
//  Sign up free at https://sms8.io (5-day trial, no card).
//

import Foundation

/// SMS8 SDK errors.
public enum SMS8Error: Error, LocalizedError {
    case missingApiKey
    case http(Int, String?)
    case decoding(Error)
    case mcp(String)
    case noData

    public var errorDescription: String? {
        switch self {
        case .missingApiKey:           return "Missing SMS8 API key. Get one at https://sms8.io"
        case .http(let code, let msg): return "HTTP \(code): \(msg ?? "")"
        case .decoding(let e):         return "Decoding error: \(e.localizedDescription)"
        case .mcp(let m):              return m
        case .noData:                  return "No response body"
        }
    }
}

/// Result of a `send` call.
public struct SendOTPResult: Codable {
    public let success: Bool
    public let otpId: Int?
    public let phone: String?
    public let expiresAt: String?
    public let expiresIn: Int?
    public let error: String?

    enum CodingKeys: String, CodingKey {
        case success
        case otpId      = "otp_id"
        case phone
        case expiresAt  = "expires_at"
        case expiresIn  = "expires_in"
        case error
    }
}

/// Result of a `verify` call.
public struct VerifyOTPResult: Codable {
    public let success: Bool
    public let verified: Bool?
    public let error: String?
    public let attemptsLeft: Int?

    enum CodingKeys: String, CodingKey {
        case success
        case verified
        case error
        case attemptsLeft = "attempts_left"
    }
}

/// SMS8 client. Sends SMS and OTP codes through your paired Android phone.
///
/// ```swift
/// let sms8 = SMS8(apiKey: "sk_xxx")
/// Task {
///     let sent = try await sms8.sendOTP(phone: "+14155550100")
///     // ...show the OTP input to the user...
///     let result = try await sms8.verifyOTP(phone: "+14155550100", code: "482937")
///     if result.verified == true { /* allow login */ }
/// }
/// ```
public final class SMS8 {
    private let apiKey: String
    private let baseURL: URL
    private let session: URLSession

    /// Initialise the client.
    /// - Parameter apiKey: Your SMS8 API key from https://app.sms8.io
    /// - Parameter baseURL: Override the MCP endpoint (rarely needed).
    /// - Parameter session: URLSession to use. Defaults to .shared.
    public init(
        apiKey: String,
        baseURL: URL = URL(string: "https://mcp.sms8.io")!,
        session: URLSession = .shared
    ) {
        self.apiKey = apiKey
        self.baseURL = baseURL
        self.session = session
    }

    // ── SMS ─────────────────────────────────────────────────────────────
    /// Send a plain SMS through your paired Android phone.
    @discardableResult
    public func sendSMS(
        phone: String,
        message: String,
        deviceId: Int? = nil,
        simSlot: String? = nil
    ) async throws -> [String: Any] {
        var args: [String: Any] = ["phone": phone, "message": message]
        if let d = deviceId { args["device_id"] = d }
        if let s = simSlot  { args["sim_slot"]  = s }
        return try await mcpCall(tool: "send_sms", arguments: args)
    }

    // ── OTP send ────────────────────────────────────────────────────────
    /// Generate and send an OTP code. Returns the otpId and expiry.
    public func sendOTP(
        phone: String,
        length: Int = 6,
        template: String? = nil,
        expiresIn: Int = 300,
        maxAttempts: Int = 5,
        deviceId: Int? = nil,
        simSlot: String? = nil
    ) async throws -> SendOTPResult {
        var args: [String: Any] = [
            "phone":        phone,
            "length":       length,
            "expires_in":   expiresIn,
            "max_attempts": maxAttempts
        ]
        if let t = template { args["template"] = t }
        if let d = deviceId { args["device_id"] = d }
        if let s = simSlot  { args["sim_slot"]  = s }
        let raw = try await mcpCall(tool: "send_otp", arguments: args)
        return try decode(raw, as: SendOTPResult.self)
    }

    // ── OTP verify ──────────────────────────────────────────────────────
    /// Verify a code the user typed. The server compares against the most
    /// recent unverified code for that phone.
    public func verifyOTP(phone: String, code: String) async throws -> VerifyOTPResult {
        let raw = try await mcpCall(tool: "verify_otp", arguments: ["phone": phone, "code": code])
        return try decode(raw, as: VerifyOTPResult.self)
    }

    // ── Internal: JSON-RPC over HTTPS ────────────────────────────────────
    private func mcpCall(tool: String, arguments: [String: Any]) async throws -> [String: Any] {
        var args = arguments
        args["api_key"] = apiKey
        let body: [String: Any] = [
            "jsonrpc": "2.0",
            "method":  "tools/call",
            "params":  ["name": tool, "arguments": args],
            "id":      Int(Date().timeIntervalSince1970 * 1000)
        ]
        var req = URLRequest(url: baseURL)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        req.setValue("sms8-swift/1.0.0", forHTTPHeaderField: "User-Agent")
        req.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse else {
            throw SMS8Error.noData
        }
        if !(200..<300).contains(http.statusCode) {
            let msg = String(data: data, encoding: .utf8)
            throw SMS8Error.http(http.statusCode, msg)
        }
        guard let env = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw SMS8Error.noData
        }
        if let err = env["error"] as? [String: Any], let m = err["message"] as? String {
            throw SMS8Error.mcp(m)
        }
        guard
            let result  = env["result"]   as? [String: Any],
            let content = result["content"] as? [[String: Any]],
            let first   = content.first,
            let text    = first["text"]   as? String
        else {
            return (env["result"] as? [String: Any]) ?? [:]
        }
        if let inner = text.data(using: .utf8),
           let parsed = try? JSONSerialization.jsonObject(with: inner) as? [String: Any] {
            return parsed
        }
        return ["text": text]
    }

    private func decode<T: Decodable>(_ dict: [String: Any], as type: T.Type) throws -> T {
        let data = try JSONSerialization.data(withJSONObject: dict)
        do {
            return try JSONDecoder().decode(type, from: data)
        } catch {
            throw SMS8Error.decoding(error)
        }
    }
}
